### PR TITLE
feat(plasma-core): usePopper in popup

### PR DESCRIPTION
**Pull request by kayman233**
Добавлена возможность автоматического позиционивания в `Popup`: `placement="auto"`
А также возможность автоматически выбирать необходимые позиции, передав массив. Например: `placement={["left", "right"]}`

Так как `Dropdown` сделан на основе `Popup`, то эти же нововведения доступны и для этого комопнента.

ezgif-1-e3814f2300

:baby_chick: Download canary assets:

[colors--canary.490.4808822080.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.490.4808822080.xml)  
[color_sberHealth_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_ios-swift--canary.490.4808822080.swift)  
[color_sberHealth_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_kotlin--canary.490.4808822080.kt)  
[color_sberHealth_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_react-native--canary.490.4808822080.ts)  
[color_sbermarket_business_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_ios-swift--canary.490.4808822080.swift)  
[color_sbermarket_business_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_kotlin--canary.490.4808822080.kt)  
[color_sbermarket_business_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_react-native--canary.490.4808822080.ts)  
[color_sbermarket_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.490.4808822080.swift)  
[color_sbermarket_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.490.4808822080.kt)  
[color_sbermarket_metro_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_ios-swift--canary.490.4808822080.swift)  
[color_sbermarket_metro_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_kotlin--canary.490.4808822080.kt)  
[color_sbermarket_metro_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_react-native--canary.490.4808822080.ts)  
[color_sbermarket_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.490.4808822080.ts)  
[color_sbermarket_selgros_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_ios-swift--canary.490.4808822080.swift)  
[color_sbermarket_selgros_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_kotlin--canary.490.4808822080.kt)  
[color_sbermarket_selgros_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_react-native--canary.490.4808822080.ts)  
[color_sbermarket_wlbusiness_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_wlbusiness_ios-swift--canary.490.4808822080.swift)  
[color_sbermarket_wlbusiness_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_wlbusiness_kotlin--canary.490.4808822080.kt)  
[color_sbermarket_wlbusiness_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_wlbusiness_react-native--canary.490.4808822080.ts)  
[color_sberprime_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.490.4808822080.swift)  
[color_sberprime_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.490.4808822080.kt)  
[color_sberprime_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.490.4808822080.ts)  
[PlasmaTokensColor--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.490.4808822080.swift)  
[shadow_sbermarket_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/shadow_sbermarket_react-native--canary.490.4808822080.ts)  
[typo_mage_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.490.4808822080.swift)  
[typo_mage_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.490.4808822080.kt)  
[typo_mage_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.490.4808822080.ts)  
[typo_plasma_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.490.4808822080.swift)  
[typo_plasma_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.490.4808822080.kt)  
[typo_plasma_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.490.4808822080.ts)  
[typo_ruler_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.490.4808822080.swift)  
[typo_ruler_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.490.4808822080.kt)  
[typo_ruler_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.490.4808822080.ts)  
[typo_sage_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.490.4808822080.swift)  
[typo_sage_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.490.4808822080.kt)  
[typo_sage_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.490.4808822080.ts)  
[typo_sbermarket_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.490.4808822080.swift)  
[typo_sbermarket_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.490.4808822080.kt)  
[typo_sbermarket_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.490.4808822080.ts)  
[typo_soulmate_ios-swift--canary.490.4808822080.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.490.4808822080.swift)  
[typo_soulmate_kotlin--canary.490.4808822080.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.490.4808822080.kt)  
[typo_soulmate_react-native--canary.490.4808822080.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.490.4808822080.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.205.0-canary.490.4808822080.0
  npm install @salutejs/plasma-core@1.112.0-canary.490.4808822080.0
  npm install @salutejs/plasma-hope@1.205.0-canary.490.4808822080.0
  npm install @salutejs/plasma-icons@1.141.0-canary.490.4808822080.0
  npm install @salutejs/plasma-temple@1.155.0-canary.490.4808822080.0
  npm install @salutejs/plasma-ui@1.187.0-canary.490.4808822080.0
  npm install @salutejs/plasma-web@1.205.0-canary.490.4808822080.0
  npm install @salutejs/plasma-sb-utils@0.110.0-canary.490.4808822080.0
  # or 
  yarn add @salutejs/plasma-b2c@1.205.0-canary.490.4808822080.0
  yarn add @salutejs/plasma-core@1.112.0-canary.490.4808822080.0
  yarn add @salutejs/plasma-hope@1.205.0-canary.490.4808822080.0
  yarn add @salutejs/plasma-icons@1.141.0-canary.490.4808822080.0
  yarn add @salutejs/plasma-temple@1.155.0-canary.490.4808822080.0
  yarn add @salutejs/plasma-ui@1.187.0-canary.490.4808822080.0
  yarn add @salutejs/plasma-web@1.205.0-canary.490.4808822080.0
  yarn add @salutejs/plasma-sb-utils@0.110.0-canary.490.4808822080.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-490/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-490/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-490/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-490/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-490/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-490/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-490/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-490/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-490/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-490/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-490/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-490/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-490/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-490/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-490/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Review by Yakutoc — APPROVED** [review]

File: packages/plasma-core/api/plasma-core.api.md [hunk 1]
```diff
@@ -814,7 +814,7 @@ export interface PopupProps extends HTMLAttributes<HTMLDivElement> {
     disclosure?: ReactNode;
     isOpen?: boolean;
     onToggle?: (isOpen: boolean, event: SyntheticEvent | Event) => void;
-    placement: 'bottom' | 'right' | 'left';
+    placement?: 'top' | 'bottom' | 'right' | 'left' | 'auto';
```
Comment (Yakutoc) [inline packages/plasma-core/api/plasma-core.api.md:817]: @kayman233 У нас есть возможность написать cypress тест на это?

Comment (Yeti-or) [reply packages/plasma-core/api/plasma-core.api.md:817]: нам точно нужно тесты добавить да

File: packages/plasma-core/src/components/Popup/Popup.tsx [hunk 2]
```diff
@@ -143,6 +128,14 @@ export const Popup = memo<PopupProps & RefAttributes<HTMLDivElement>>(
                 return () => document.removeEventListener('click', onDocumentClick);
             }, []);
 
+            useEffect(() => {
+                if (!isOpen || !forceUpdate) {
```
Comment (Yeti-or) [inline packages/plasma-core/src/components/Popup/Popup.tsx:132]: чет у нас где `isVisible` где `isOpen` надо задачу завести на причесать это в plasma-2.0, заведи плиз

File: packages/plasma-core/src/components/Popup/Popup.tsx [hunk 3]
```diff
@@ -71,11 +52,15 @@ const StyledPopup = styled.div<Pick<PopupProps, 'placement'>>`
...
 export const Popup = memo<PopupProps & RefAttributes<HTMLDivElement>>(
     React.forwardRef<HTMLDivElement, PopupProps>(
-        ({ disclosure, children, isOpen, trigger, placement, onToggle, ...rest }, outerRootRef) => {
+        ({ disclosure, children, isOpen, trigger, placement = 'auto', onToggle, ...rest }, outerRootRef) => {
             const rootRef = useRef<HTMLDivElement | null>(null);
             const popupRef = useRef<HTMLDivElement | null>(null);
             const handleRef = useForkRef<HTMLDivElement>(rootRef, outerRootRef);
 
+            const { styles, attributes, forceUpdate } = usePopper(rootRef.current, popupRef.current, {
+                placement: `${placement}-start` as PopperJS.Placement,
```
Comment (neretin-trike) [inline packages/plasma-core/src/components/Popup/Popup.tsx:61]: а напомни пожалуйста, зачем в конце `start` добавлять? Попер принимает и обычные значения https://popper.js.org/docs/v2/constructors/#options

Comment (kayman233) [reply packages/plasma-core/src/components/Popup/Popup.tsx:61]: чтобы обратной совместимости добиться. У нас изначально в `Popup` было свое расположение, поведение которого через `react-popper`, можно добиться только добавляя `start` к исходным

Comment (neretin-trike) [reply packages/plasma-core/src/components/Popup/Popup.tsx:61]: уловил, спасиб 👍

File: packages/plasma-core/src/components/Popup/Popup.tsx [hunk 4]
```diff
@@ -16,7 +18,7 @@ export interface PopupProps extends HTMLAttributes<HTMLDivElement> {
     /**
      * Расположение всплывающего окна.
```
Comment (Yeti-or) [inline packages/plasma-core/src/components/Popup/Popup.tsx:19]: какое положение по умолчанию

File: packages/plasma-core/src/components/Popup/Popup.tsx [hunk 5]
```diff
@@ -156,9 +149,9 @@ export const Popup = memo<PopupProps & RefAttributes<HTMLDivElement>>(
                     {disclosure}
                     {children && (
                         <StyledPopup
+                            {...attributes.popper}
                             ref={popupRef}
-                            placement={placement}
-                            style={{ display: isOpen ? 'block' : 'none' }}
+                            style={{ ...styles.popper, ...{ display: isOpen ? 'block' : 'none' } }}
```
Comment (Yeti-or) [inline packages/plasma-core/src/components/Popup/Popup.tsx:154]: а почему у нас тут тупо не на уровне jsx вырубается popup?
из-за анимации?

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-490/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-490/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-490/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-490/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-490/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-490/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-490/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-490/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-490/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-490/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-490/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-490/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-490/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-490/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

File: packages/plasma-core/src/components/Popup/Popup.tsx (hunk 5 above)
Comment (kayman233) [reply packages/plasma-core/src/components/Popup/Popup.tsx:181]: если на уровне jsx вырубать, то поведение странное - попап появляется в разных местах. Видимо, самому popper-у, чтобы оба элемента были замаучены(ну собственно я ему рефки их и передаю)

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-490/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-490/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-490/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-490/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

File: packages/plasma-core/src/components/Popup/Popup.tsx [hunk 6]
```diff
@@ -14,9 +19,13 @@ export interface PopupProps extends HTMLAttributes<HTMLDivElement> {
      */
     trigger: 'hover' | 'click';
     /**
-     * Расположение всплывающего окна.
+     * Расположение всплывающего окна. По умолчанию "auto"
+     */
+    placement?: PopupPlacement;
```
Comment (Yeti-or) [inline packages/plasma-core/src/components/Popup/Popup.tsx:24]:
```js
placement? PopupPlacement | Array<PopupBasicPlacement>;

const auto = Array.isArray(placement);
```

File: website/plasma-web-docs/docs/components/Dropdown.mdx [hunk 7]
```diff
@@ -267,6 +267,50 @@ export function App() {
 }
 ```
 
+## Позицирование
+Помимо стандартных способов расположения, можно также указать значение `auto`, которое будет автоматически определять нужное расположение.
+
+Также при использовании `auto` можно явно указать, в каких направлениях можно открывать выпадающий список с помощью `allowedAutoPlacements`.
```
Comment (Yeti-or) [inline website/plasma-web-docs/docs/components/Dropdown.mdx:273]: и тут поменять

**Review by Yeti-or — COMMENTED** [review]
Давай добавим в Storybook

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-490/

**Review by Yeti-or — APPROVED** [review]
Только в шапке pr обнови изменения

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-490/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-490/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-490/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-490/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-490/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-490/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Pull request merged**